plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -0,0 +1,375 @@
+      applyHints' (Just (Left err)) = Left err
+      applyHints' Nothing = Right []
+
+  fmap applyHints' (moduleEx flags)
+
+  where moduleEx :: ParseFlags -> Action (Maybe (Either ParseError ModuleEx))
+#ifdef GHC_LIB
+        moduleEx flags = do
+          flags' <- setExtensions flags
+          (_, contents) <- getFileContents nfp
+          let fp = fromNormalizedFilePath nfp
+          let contents' = T.unpack <$> contents
+          Just <$> (liftIO $ parseModuleEx flags' fp contents')
+
+        setExtensions flags = do
+          hsc <- hscEnv <$> use_ GhcSession nfp
+          let dflags = hsc_dflags hsc
+          let hscExts = EnumSet.toList (extensionFlags dflags)
+          let hlintExts = mapMaybe (GhclibParserEx.readExtension . show) hscExts
+          logm $ "hlint:getIdeas:setExtensions:" ++ show hlintExts
+          return $ flags { enabledExtensions = hlintExts }
+#else
+        moduleEx _flags = do
+          mbpm <- getParsedModule nfp
+          return $ createModule <$> mbpm
+          where createModule pm = Right (createModuleEx anns modu)
+                  where anns = pm_annotations pm
+                        modu = pm_parsed_source pm
+#endif
+
+-- ---------------------------------------------------------------------
+
+data HlintUsage
+  = HlintEnabled { cmdArgs :: [String] }
+  | HlintDisabled
+  deriving Show
+
+data GetHlintSettings = GetHlintSettings
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable GetHlintSettings
+instance NFData   GetHlintSettings
+instance NFData Hint where rnf = rwhnf
+instance NFData Classify where rnf = rwhnf
+instance NFData ParseFlags where rnf = rwhnf
+instance Show Hint where show = const "<hint>"
+instance Show ParseFlags where show = const "<parseFlags>"
+instance Binary GetHlintSettings
+
+type instance RuleResult GetHlintSettings = (ParseFlags, [Classify], Hint)
+
+getHlintSettingsRule :: HlintUsage -> Rules ()
+getHlintSettingsRule usage =
+    defineNoFile $ \GetHlintSettings ->
+      liftIO $ case usage of
+          HlintEnabled cmdArgs -> argsSettings cmdArgs
+          HlintDisabled -> fail "hlint configuration unspecified"
+
+-- ---------------------------------------------------------------------
+
+codeActionProvider :: CodeActionProvider
+codeActionProvider _lf ideState plId docId _ context = Right . LSP.List . map CACodeAction <$> getCodeActions
+  where
+
+    getCodeActions = do
+        applyOne <- applyOneActions
+        diags <- getDiagnostics ideState
+        let docNfp = toNormalizedFilePath' <$> uriToFilePath' (docId ^. LSP.uri)
+            numHintsInDoc = length
+              [d | (nfp, _, d) <- diags
+                 , d ^. LSP.source == Just "hlint"
+                 , Just nfp == docNfp
+              ]
+        -- We only want to show the applyAll code action if there is more than 1
+        -- hint in the current document
+        if numHintsInDoc >= 2 then do
```

Comment:
Why not > 1 - the comment talks about greater than 1, but the code talks about 2.

Reply:
Should we be considering how many hints have suggestions?

Reply:
conditional changed to match the comment and number of hints to use `validCommand` (that already take in account the hint has suggestion)